parse.y

```diff
@@ -6873,7 +6890,11 @@ parser_here_document(struct parser_params *parser, NODE *here)
 			    yylval.val, str);
 #endif
     heredoc_restore(lex_strterm);
-    lex_strterm = NEW_STRTERM(-1, 0, 0);
+    if (func & STR_FUNC_REGEXP) {
+	lex_strterm = NEW_STRTERM(-2, 0, 0);
+    } else {
+	lex_strterm = NEW_STRTERM(-1, 0, 0);
+    }
```

Comment:
This is the exit point for the non '~' cases.  It seems this hack is here so we can return 2 tokens.  tSTRING_CONTENT and (tREGEXP_END|tSTRING_END).  But I would think there is a better way.